### PR TITLE
[now-next] Add headers support for custom-routes

### PR DESCRIPTION
**Pull request by ijjk**
This adds support for `headers` in custom-routes which was landed in Next.js. 

This also updates `@now/routing-utils` `convertHeaders` to call `sourceToRegex` to match behavior with Next.js and allow using `segments` to match in the header `source` as not being able to use the same syntax for a header `source` as a `redirect` source could get confusing

### CODE REVIEW

**Review by dav-is — APPROVED** [review]
LGTM

File: packages/now-routing-utils/src/superstatic.ts [hunk 1]
```diff
@@ -75,7 +75,7 @@ export function convertHeaders(headers: NowHeader[]): Route[] {
       obj[kv.key] = kv.value;
     });
     const route: Route = {
-      src: h.source,
+      src: sourceToRegex(h.source).src,
```
Comment (styfle) [inline packages/now-routing-utils/src/superstatic.ts]: This doesn't seem right. This will allow segments in headers? We didn't spec this.

**Review by styfle — DISMISSED** [review]
Blocking until we can discuss further.

We need to land #3537 before merging any other changes to routing properties.

File: packages/now-routing-utils/src/superstatic.ts (hunk 1 above)
Comment (dav-is) [reply packages/now-routing-utils/src/superstatic.ts]: Seems reasonable to me

Say you want to add cache headers to `/blog/:post` you don't want to have to write `/blog/([^/]+)`

File: packages/now-next/src/utils.ts [hunk 2]
```diff
@@ -302,6 +302,14 @@ export type Redirect = Rewrite & {
   statusCode?: number;
 };
 
+export type Header = {
```
Comment (styfle) [inline packages/now-next/src/utils.ts]: This type (and many others above) are available in routing utils:

```js
import { NowHeader } from `@now/routing-utils`;
```

File: packages/now-routing-utils/test/superstatic.spec.js [hunk 3]
```diff
@@ -412,12 +412,12 @@ test('convertHeaders', () => {
...
+      src: '^(.*)+(?:\\/(.*))\\.(eot|otf|ttf|ttc|woff|font\\.css)$',
       headers: { 'Access-Control-Allow-Origin': '*' },
       continue: true,
     },
     {
-      src: '404.html',
+      src: '^404\\.html$',
       headers: { 'Cache-Control': 'max-age=300', 'Set-Cookie': 'error=404' },
       continue: true,
     },
```
Comment (styfle) [inline packages/now-routing-utils/test/superstatic.spec.js:436]: We need a new test to prove that segments in `headers` works properly. You can another route in this test actually.

File: packages/now-routing-utils/src/superstatic.ts [hunk 4]
```diff
@@ -75,7 +75,7 @@ export function convertHeaders(headers: NowHeader[]): Route[] {
...
       headers: obj,
       continue: true,
     };
```
Comment (styfle) [inline packages/now-routing-utils/src/superstatic.ts:99]: We need to validate headers in `index.ts`.

Search for:

```
Redirect `source` contains invalid regex.
```

We need a similar validation for `headers` too.

Comment (ijjk) [reply packages/now-routing-utils/src/superstatic.ts:99]: Added a couple tests for this

File: packages/now-routing-utils/test/superstatic.spec.js (hunk 3 above)
Comment (ijjk) [reply packages/now-routing-utils/test/superstatic.spec.js:436]: Added another test for segments in the header source

File: packages/now-next/src/utils.ts (hunk 2 above)
Comment (ijjk) [reply packages/now-next/src/utils.ts]: Updated to re-use some types from `@now/routing-utils`

File: packages/now-next/src/index.ts [hunk 5]
```diff
@@ -990,6 +1003,16 @@ export const build = async ({
...
+      Desired routes order
+      - Runtime headers
+      - User headers and redirects
+      - Runtime redirects
+      - Runtime routes
+      - Check filesystem, if nothing found continue
+      - User rewrites
+      - Builder rewrites
+    */
     routes: [
```
Comment (dav-is) [inline packages/now-next/src/index.ts:1014]: Do you need to add `headers` here too?

Comment (ijjk) [reply packages/now-next/src/index.ts:1014]: Yeah, looks like they got removed when resolving merge conflict 😅

File: packages/now-next/src/utils.ts [hunk 6]
```diff
@@ -12,8 +12,9 @@ import {
   streamToBuffer,
   Lambda,
   isSymbolicLink,
+  NowRewrite,
```
Comment (styfle) [inline packages/now-next/src/utils.ts]: This is strange... `NowRewrite` should be imported from `@now/routing-utils`

File: packages/now-routing-utils/test/superstatic.spec.js [hunk 7]
```diff
@@ -408,31 +408,49 @@ test('convertHeaders', () => {
         },
       ],
     },
+    {
+      source: '/blog/:path*',
+      headers: [
+        {
+          key: 'on-blog',
+          value: 'yup',
+        },
```
Comment (styfle) [inline packages/now-routing-utils/test/superstatic.spec.js:421]: @ijjk @dav-is Do we want to support path segments for key/value?

For example:
```js
{
  source: '/blog/:post',
  headers: [{
    key: 'x-blog-post',
    value: ':post',
  }]
}
```

Comment (dav-is) [reply packages/now-routing-utils/test/superstatic.spec.js:421]: I think `$1` would still work so might as well

Comment (ijjk) [reply packages/now-routing-utils/test/superstatic.spec.js:421]: Double checked `$1` works so makes sense for segments to work too 👍

**Review by styfle — APPROVED** [review]
Nice!

File: packages/now-next/src/utils.ts (hunk 6 above)
Comment (ijjk) [reply packages/now-next/src/utils.ts]: Oh interesting, must of auto-imported from there instead. Updated

**Pull request merged**